File: creational/prototype.ts
// In each implementation, the Shape interface defines the clone and draw methods.
// The Circle and Square classes implement the Shape interface.
// The PrototypeFactory class is responsible for managing and cloning prototypes.
// This setup allows for the creation of new shape instances by cloning existing prototypes,
// demonstrating the Prototype design pattern.

// Shape interface
interface Shape {
  clone(): Shape;
  draw(): void;
}

// Circle class
class Circle implements Shape {
  constructor(private radius: number) {}

  clone(): Circle {
    return new Circle(this.radius);
  }

  draw(): void {
    console.log(`Drawing Circle with radius: ${this.radius}`);
  }
}

// Square class
class Square implements Shape {
  constructor(private side: number) {}

  clone(): Square {
    return new Square(this.side);
  }

  draw(): void {
    console.log(`Drawing Square with side: ${this.side}`);
  }
}

// Prototype Factory
class PrototypeFactory {
  private prototypes: Map<string, Shape> = new Map();

  registerPrototype(name: string, prototype: Shape): void {
    this.prototypes.set(name, prototype);
  }

  createPrototype(name: string): Shape {
    const prototype = this.prototypes.get(name);
    if (!prototype) {
      throw new Error(`Prototype not found: ${name}`);
    }
    return prototype.clone();
  }
}

// Example usage
const factory = new PrototypeFactory();

factory.registerPrototype("circle", new Circle(10));
factory.registerPrototype("square", new Square(5));

const circle = factory.createPrototype("circle");
const square = factory.createPrototype("square");

circle.draw();
square.draw();

File: creational/prototype.py
from abc import ABC, abstractmethod
from typing import Dict


# Shape interface
class Shape(ABC):
    @abstractmethod
    def clone(self) -> "Shape":
        pass

    @abstractmethod
    def draw(self) -> None:
        pass


# Circle class
class Circle(Shape):
    def __init__(self, radius: int) -> None:
        self.radius = radius

    def clone(self) -> "Circle":
        return Circle(self.radius)

    def draw(self) -> None:
        print(f"Drawing Circle with radius: {self.radius}")


# Square class
class Square(Shape):
    def __init__(self, side: int) -> None:
        self.side = side

    def clone(self) -> "Square":
        return Square(self.side)

    def draw(self) -> None:
        print(f"Drawing Square with side: {self.side}")


# Prototype Factory
class PrototypeFactory:
    def __init__(self) -> None:
        self._prototypes: Dict[str, Shape] = {}

    def register_prototype(self, name: str, prototype: Shape) -> None:
        self._prototypes[name] = prototype

    def create_prototype(self, name: str) -> Shape:
        prototype = self._prototypes.get(name)
        if prototype is None:
            raise ValueError(f"Prototype not found: {name}")
        return prototype.clone()


# Example usage
factory = PrototypeFactory()

factory.register_prototype("circle", Circle(10))
factory.register_prototype("square", Square(5))

circle = factory.create_prototype("circle")
square = factory.create_prototype("square")

circle.draw()
square.draw()
